contract-message-predicate/src/predicate_asm.rs

```diff
@@ -0,0 +1,76 @@
+    const INPUT_TYPE: u8 = 0x18;
+    const INPUT_MSG_DATA_LEN: u8 = 0x19;
+    const EXPECTED_INPUT_TYPE: u8 = 0x20;
+
+    /* The following assembly code is intended to do the following:
+     *  -Verify that the script bytecode hash for the transaction matches that of
+     *   the expected Message to Contract script
+     *  -Verify there are no other `InputMessages` with data in the transaction
+     *   other than the first input
+     *
+     * If these conditions are met, then the predicate evaluates as true.
+     */
+    let mut predicate: Vec<u8> = vec![
+        //extend stack for storing script hash
+        op::move_(SCRIPT_HASH_PTR, STACK_PTR), //SCRIPT_HASH_PTR = stack pointer
+        op::cfei(32),                          //extends current call frame stack by 32 bytes
+        //compute script hash
+        op::gtf(SCRIPT_PTR, ZERO, GTF_SCRIPT), //SCRIPT_PTR = script data address
+        op::gtf(SCRIPT_LEN, ZERO, GTF_SCRIPT_LEN), //SCRIPT_LEN = script data length
+        op::s256(SCRIPT_HASH_PTR, SCRIPT_PTR, SCRIPT_LEN), //32bytes at SCRIPT_HASH_PTR = hash of the script
+        //compare hash with expected
+        op::addi(EXPECTED_HASH_PTR, INSTR_START, 19 * BYTES_PER_INSTR), //EXPECTED_HASH_PTR = address of reference data at end of program
+        op::movi(VAL_32, 32),                                           //VAL_32 = 32
+        op::meq(COMPARE_RESULT, EXPECTED_HASH_PTR, SCRIPT_HASH_PTR, VAL_32), //COMPARE_RESULT = if the 32bytes at SCRIPT_HASH_PTR equals the 32bytes at EXPECTED_HASH_PTR
+        op::jnei(COMPARE_RESULT, ONE, 18), //jumps to PREDICATE_FAILURE if COMPARE_RESULT is not 1
+        //confirm that no other messages with data are included
+        op::gtf(INPUT_INDEX, ZERO, GTF_SCRIPT_INPUTS_COUNT), //INPUT_INDEX = the number of inputs in the script
+        op::addi(EXPECTED_INPUT_TYPE, ZERO, INPUT_MESSAGE_TYPE), //EXPECTED_INPUT_TYPE = INPUT_MESSAGE_TYPE
+        //LOOP_START:
+        op::subi(INPUT_INDEX, INPUT_INDEX, 1), //INPUT_INDEX = INPUT_INDEX - 1
+        //check if the input is a message input
+        op::gtf(INPUT_TYPE, INPUT_INDEX, GTF_INPUT_TYPE), //INPUT_TYPE = the type of input for input[INPUT_INDEX]
+        op::jnei(INPUT_TYPE, EXPECTED_INPUT_TYPE, 16), //skips to SKIP_DATA_CHECK if INPUT_TYPE does not equal EXPECTED_INPUT_TYPE
+        //check it the input message has data
+        op::gtf(INPUT_MSG_DATA_LEN, INPUT_INDEX, GTF_MSG_DATA_LEN), //INPUT_MSG_DATA_LEN = the data length of input[INPUT_INDEX]
+        op::jnei(INPUT_MSG_DATA_LEN, ZERO, 18), //jumps to PREDICATE_FAILURE if INPUT_MSG_DATA_LEN does not equal 0
+        //SKIP_DATA_CHECK:
+        op::jnei(INPUT_INDEX, ONE, 11), //jumps back to LOOP_START if INPUT_INDEX does not equal 1
+        op::ret(ONE),
+        //PREDICATE_FAILURE:
+        op::ret(ZERO),
+        //referenced data (expected script hash)
+        //00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
+    ]
+    .into_iter()
```

Comment:
Great thinking with using the heap vs the stack!

Reply:
I spoke too soon... there is no immediate version of `ALOC` like there is `CFEI` so it would require at least 2 instructions to increase the heap.

Reply:
I moved the jump instruction values to constants since some of them are referenced more than once. We could add a check for things like the value for the end of the program code before adding referenced data bits, but we effectively already are indirectly through the tests (particularly the test where everything has to work correctly).

Reply:
>once the new ALOC semantics are merged
Using the heap may be an optimization that has to wait, not sure when it will land.

Using `jnzi` sounds like an easy win though.
I don't have anything to add; @otrho was too thorough :)

Reply:
> I spoke too soon... there is no immediate version of `ALOC` like there is `CFEI` so it would require at least 2 instructions to increase the heap.

Ahh, yeah.  Bugger.  OK, I think it has been golfed all the way.